src/module.c

```diff
@@ -5347,9 +5352,21 @@ int VM_HashSet(ValkeyModuleKey *key, int flags, ...) {
          * to avoid a useless copy. */
         if (flags & VALKEYMODULE_HASH_CFIELDS) low_flags |= HASH_SET_TAKE_FIELD;
 
-        robj *argv[2] = {field, value};
-        hashTypeTryConversion(key->value, argv, 0, 1);
-        int updated = hashTypeSet(key->value, field->ptr, value->ptr, low_flags);
+        char *value_sds;
+        if (flags & VALKEYMODULE_HASH_SHAREBLE_VALUES) {
+            if (key->value->encoding == OBJ_ENCODING_LISTPACK) {
+                /* Convert to hashtable encoding, as list pack encoding performs a deep copy
+                 * of the buffer, breaking ref-counting semantics. */
+                hashTypeConvert(key->value, OBJ_ENCODING_HASHTABLE);
+            }
+            value_sds = ((ValkeyModuleSharedSDS *)value)->buf;
+        } else {
+            value_sds = value->ptr;
+            robj *argv[2] = {field, value};
+            hashTypeTryConversion(key->value, argv, 0, 1);
+        }
+
+        int updated = hashTypeSet(key->value, field->ptr, value_sds, low_flags);
```

Comment:
This code has some technical dept already and here we're adding more.

hashTypeSet already converts to hashtable if needed, so hashTypeTryConversion wasn't needed here before and it isn't needed here how, unless I'm missing something.

I think we should move the handling of SDS_TYPE_32_SHARED to hashTypeSet and convert to hashtable there, and not have this logic here in module.c. This stuff belongs better in the hash type code.

```suggestion
        sds value_sds;
        if (flags & VALKEYMODULE_HASH_SHAREBLE_VALUES) {
            value_sds = ((ValkeyModuleSharedSDS *)value)->buf;
        } else {
            value_sds = value->ptr;
        }

        int updated = hashTypeSet(key->value, field->ptr, value_sds, low_flags);
```

Reply:
This makes sense